### PR TITLE
Update React v0.13.0 definitions for RC2

### LINKED ISSUE #3795: React 0.13 cloneElement
I stumbled upon fact that React.cloneElement is not defined. I played a bit with definition and come up with
```
    function cloneElement<P>(
        element:ReactElement<P>, newProps?: P, ...children: ReactNode[]): ReactElement<P>;
    function cloneElement<P>(
        element:ReactElement<P>, config?: any, ...children: ReactNode[]): ReactElement<P>;
```

this is wrong and incomplete :) I just want for someone to make it better :)

### PR DESCRIPTION
**Pull request by vsiao**
Added:
- `React.cloneElement` (fixes #3795)
- `React.addons.createFragment`
- `setState(f: (prevState: S, props: P) => S, callback?: () => any)` (transactional `setState`)
- `ref: (component: T) => any` (callback refs)

It's common to reference the type of generic React elements. With the previous type definitions, `React.ReactElement<P>` specifically referred to "modern" elements created from modern component classes, which is confusing. So now we have:

changed: `React.ReactElementBase<T, P>` to `React.ReactElement<P>` (added `React.ModernElement<P>`)
changed: `React.ComponentClassBase<P>` to `React.ComponentClass<P>` (added `React.ModernComponentClass<P, S>`)

With callback refs, we want to be able to infer the type of the component passed in the callback. To do so, extend `React.Props<T>` like so:

``` ts
interface MyProps extends React.Props<MyComponent> {}
class MyComponent extends React.Component<MyProps, {}> {}

React.createElement(MyComponent, { ref: (component: MyComponent) => null }); // ok
React.createElement(MyComponent, { ref: (component: React.DOMComponent) => null }); // fail
```

cc @pspeter3, @jbrantly

### CODE REVIEW

File: react/future/react-0.13.0.d.ts [hunk 1]
```diff
@@ -8,24 +8,33 @@ declare module "react" {
     // React Elements
     // ----------------------------------------------------------------------
 
-    interface ReactElementBase<T, P> {
-        type: T;
+    type ReactType = ComponentClass<any> | string;
+
+    interface ReactElement<P> {
+        type: string | ComponentClass<P>;
```
Comment (pspeter3) [inline react/future/react-0.13.0.d.ts]: I like this cleanup

File: react/future/react-0.13.0.d.ts [hunk 2]
```diff
@@ -8,24 +8,33 @@ declare module "react" {
...
-    interface ReactClassicElement<P>
-        extends ReactElementBase<ClassicComponentClass<P, any> | string, P> {}
+    interface ClassicElement<P> extends ReactElement<P> {
+        type: string | ClassicComponentClass<P, any>;
+        ref: string | ((component: ClassicComponent<P, any>) => any);
+    }
 
-    interface ReactDOMElement<P> // subtype of ReactClassicElement
-        extends ReactElementBase<string, P> {}
+    // subtype of ClassicElement
```
Comment (pspeter3) [inline react/future/react-0.13.0.d.ts]: I think we can explicitly extend ClassicElement

File: react/future/react-0.13.0.d.ts [hunk 3]
```diff
@@ -155,18 +183,18 @@ declare module "react" {
...
+    interface ComponentClass<P> {
         propTypes?: ValidationMap<P>;
         contextTypes?: ValidationMap<any>;
         childContextTypes?: ValidationMap<any>;
     }
 
-    interface ComponentClass<P, S> extends ComponentClassBase<P> {
+    interface ModernComponentClass<P, S> extends ComponentClass<P> {
         new(props?: P, context?: any): Component<P, S>;
         defaultProps?: P;
```
Comment (pspeter3) [inline react/future/react-0.13.0.d.ts]: Can this not also have a displayName?

File: react/future/react-0.13.0.d.ts [hunk 4]
```diff
@@ -311,13 +339,13 @@ declare module "react" {
     // Props / DOM Attributes
     // ----------------------------------------------------------------------
 
-    interface Props {
+    interface Props<T> {
         children?: ReactNode;
-        key?: number | string;
-        ref?: string;
+        key?: string | number;
+        ref?: string | ((component: T) => any);
```
Comment (pspeter3) [inline react/future/react-0.13.0.d.ts]: Cool!

**Comment by pspeter3** [comment]
This looks good to me and seems small in scope. Unless @jbrantly has objections, I think this can be merged in. We'll probably need to do a final one for React 0.13 real

File: react/future/react-0.13.0.d.ts (hunk 2 above)
Comment (vsiao) [inline react/future/react-0.13.0.d.ts]: oh, yeah.

File: react/future/react-0.13.0.d.ts (hunk 3 above)
Comment (vsiao) [inline react/future/react-0.13.0.d.ts]: You can give it a `displayName`, but it's not provided by the framework

Comment (pspeter3) [inline react/future/react-0.13.0.d.ts]: Should we make it optional then?

File: react/future/react-0.13.0.d.ts [hunk 5]
```diff
@@ -8,24 +8,33 @@ declare module "react" {
...
-        ref: string;
+        key: string | number;
+        ref: string | ((component: Component<P, any>) => any);
     }
 
-    interface ReactElement<P>
-        extends ReactElementBase<ComponentClass<P, any>, P> {}
+    interface ModernElement<P> extends ReactElement<P> {
+        type: ModernComponentClass<P, any>;
+        ref: string | ((component: Component<P, any>) => any);
```
Comment (jbrantly) [inline react/future/react-0.13.0.d.ts]: The `ref` property should already been defined from ReactElement.

**Comment by jbrantly** [comment]
It took me a minute to wrap my head around what you're trying to accomplish with the ReactElement and ComponentClass renaming stuff. You mentioned that `ReactElement` referred specifically to modern classes. While that was sort of true, I think it's mostly because the way React is structuring classic-vs-modern is that modern is a subset of classic. In other words, modern can be used as a base to extend classic from. I'm personally good with getting rid of `ReactElementBase`, but I wonder if we shouldn't also get rid of `ModernElement` and `ModernComponentClass`. I feel that these can be merged into `ReactElement` and `ComponentClass`.

For example:

```
    interface ReactElement<P> {
        type: string | ComponentClass<P>;
        props: P;
        key: string | number;
        ref: string | ((component: Component<P, any>) => any);
    }

    interface ClassicElement<P> extends ReactElement<P> {
        type: string | ClassicComponentClass<P, any>;
        ref: string | ((component: ClassicComponent<P, any>) => any);
    }

    interface ComponentClass<P> {
        new(props?: P, context?: any): Component<P, S>;
        propTypes?: ValidationMap<P>;
        contextTypes?: ValidationMap<any>;
        childContextTypes?: ValidationMap<any>;
        defaultProps?: P;
    }

    interface ClassicComponentClass<P, S> extends ComponentClass<P> {
        new(props?: P, context?: any): ClassicComponent<P, S>;
        getDefaultProps?(): P;
        displayName?: string;
    }
```

**Comment by jbrantly** [comment]
Looked over the remaining changes. Man having duplicates suck (I'm sure on the authoring side as well). I'm good with the remaining changes but I'd definitely like some conversation on the whole splitting-out-of-modern thing as that had ripple effects on the rest of the codebase as well (ex: all of the top-level APIs like `createFactory`, etc). I personally don't think we need to split common-vs-modern-vs-classic throughout the codebase, and instead think we should do common/modern-vs-classic.

**Comment by pspeter3** [comment]
Let's get rid of the Modern / Classic distinction as much as possible. It seems like it will simplify things greatly.

**Comment by vsiao** [comment]
That's a great suggestion @jbrantly—I've updated the PR.

**Comment by jbrantly** [comment]
+1 to merge

**Comment by pspeter3** [comment]
cc @vvakame

**Comment by pspeter3** [comment]
@jbrantly I've made the definitions the default. cc @vvakame @johnnyreilly

**Comment by jbrantly** [comment]
I can't comment on the changes to the other libs, but for making 0.13 the default +1 from me.

**Comment by pspeter3** [comment]
The changes to the other libs are just to have them point at the legacy
file for backwards compatibility. My thought was that they can update
individually at their discretion.

**Comment by johnnyreilly** [comment]
Thanks!

**Pull request merged**